### PR TITLE
Fix transaction route

### LINKED ISSUE #2115: Fix transaction route
**Summary or problem description**
This is an issue similar to #1667. Currently incoming transactions will be handled in the following steps:

(1) Deserialized by `RemoteNode`

(2) Sent to `Blockchain`, check whether it's within `mempool`

(3) Sent to `TransactionRouter `to check witness

(4) Sent to `Blockchain` again to check fee, whether it's within `mempool`, and added to `mempool`

There are 2 porblems:

(1) There is a circled stream from step (2) to (4). `Blockchain `is called twice which would cause TPS drop.

(2) Func `ContainsTransaction` is called twice for a single transaction in step (2) and (4)

**Do you have any solution you want to propose?**

A new scenario for incoming transactions:

(1) Deserialized by `RemoteNode`

(2) Sent to `TransactionRouter `to check witness

(3) Sent to `Blockchain` to check fee, whether it's within `mempool`, and added to `mempool`

**Neo Version**
- Neo 3

### PR DESCRIPTION
**Pull request by Qiao-Jin**
Close #2115

### CODE REVIEW

File: src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs [hunk 1]
```diff
@@ -294,15 +294,19 @@ private void OnInventoryReceived(IInventory inventory)
...
                 case Block block:
                     if (block.Index > Blockchain.Singleton.Height + InvPayload.MaxHashesCount) return;
                     UpdateLastBlockIndex(block.Index, false);
+                    system.Blockchain.Tell(inventory);
+                    break;
+                default:
+                    system.Blockchain.Tell(inventory);
                     break;
             }
             knownHashes.Add(inventory.Hash);
```
Comment (shargon) [inline src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs:308]: I think that we should move this line to the top

Comment (Qiao-Jin) [reply src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs]: have fixed

**Comment by erikzhang** [comment]
We need a unified inventory message stream. If the original stream does not have a great impact on TPS, I think there is no need to modify it.

**Comment by shargon** [comment]
@Qiao-Jin Do you have some benchmarks?

**Comment by Qiao-Jin** [comment]
Attached is a comparison between TPS of this PR and master branch.

image

**Comment by Qiao-Jin** [comment]
@erikzhang Could you have a check about the result?

**Comment by erikzhang** [comment]
I think they are not much different.

**Comment by Qiao-Jin** [comment]
> I think they are not much different.

Not so much but still significant optimization I think.

Besides, the effect of this optimization is limited, as current bottom neck is `Blockchain `receiving transactions. This is because we have included persisiting period into the 15 s block period. This means that `Blockchain `will have no time to receive transactions if traffic is heavy. If without this limitation the effect will be more significant.

Moreover the logic `Blockchain` receiving transaction array from MemoryPool is wrong, and is corrected in this PR.

**Comment by shargon** [comment]
If the benchmarks was after the fix (https://github.com/neo-project/neo/pull/2116/commits/12c333ec50851448acd05bf5befe90f5f05f3089) this PR it's good to me

**Comment by Qiao-Jin** [comment]
Ping

**Review by shargon — APPROVED** [review]
It's good for me, wait for @erikzhang

File: src/neo/Ledger/Blockchain.cs [hunk 2]
```diff
@@ -356,12 +354,9 @@ protected override void OnReceive(object message)
                 case Block block:
                     OnInventory(block, false);
                     break;
-                case Transaction tx:
-                    OnTransaction(tx, true);
-                    break;
                 case Transaction[] transactions:
                     // This message comes from a mempool's revalidation, already relayed
-                    foreach (var tx in transactions) OnTransaction(tx, false);
+                    foreach (var tx in transactions) OnNewTransaction(tx);
```
Comment (erikzhang) [inline src/neo/Ledger/Blockchain.cs:359]: The policy changed. You need to re-verify the transactions. And if you revert this change, tps will be reduced to the same as before.

**Review by erikzhang — CHANGES_REQUESTED** [review]

File: src/neo/Ledger/Blockchain.cs (hunk 2 above)
Comment (Qiao-Jin) [reply src/neo/Ledger/Blockchain.cs:359]: I suppose Func `OnNewTransaction` will call func `VerifyStateDependent` and policy is checked there?

Comment (erikzhang) [reply src/neo/Ledger/Blockchain.cs:359]: Previously, if the policy was changed, the transactions needed to be completely re-verified, including `VerifyStateIndependent`. I'm not sure if it is unnecessary now.

Comment (Qiao-Jin) [reply src/neo/Ledger/Blockchain.cs:359]: `VerifyStateIndependent` will check only the witness correctness now. Its price will be calculated in `VerifyDependent` together with others like transfer output, etc. Besides I observed that func `VerifyStateIndependent` is untouched in #2045. Is this correct?

Comment (erikzhang) [reply src/neo/Ledger/Blockchain.cs:359]: Can you fix this in another pr?

Comment (Qiao-Jin) [reply src/neo/Ledger/Blockchain.cs:359]: #2107 should have solved this problem

Comment (erikzhang) [reply src/neo/Ledger/Blockchain.cs:359]: #2107 moves `TxRouter` to `NeoSystem` too. Can you create a pr for fixing `OnNewTransaction` only?
